### PR TITLE
docs: update runtime usage metering config

### PR DESCRIPTION
**Pull request by mashenjun**

### CODE REVIEW

**Comment by mashenjun** [comment]
@codex review

Review PR `mem9-ai/mem9#320` at current head SHA `2b6045fcea0a1d46c3b8442c4796af2c05654827`.

## Review Scope Classification

This is a `first review`.

Live PR evidence shows no previous Codex review baseline: no submitted reviews, no review threads, and no prior Codex top-level review comment were found. The only PR timeline comment observed is the Netlify deploy approval notice. Since there is no previous Codex baseline, treat "changes since the previous Codex review" as the full current PR diff from `main@4e763b6c4d2c8ba474d26ff75147971f855933db` to `shenjun/update-runtime-metering-docs@2b6045fcea0a1d46c3b8442c4796af2c05654827`.

## Stable Charter Invariants

1. Runtime usage quota and console metering use `MNEMO_RUNTIME_USAGE_*`.
2. Legacy API metering uses `MNEMO_METERING_*`.
3. Runtime usage quota and console metering do not use `MNEMO_METERING_URL`.
4. Documentation must match actual environment variable behavior in the codebase.
5. Documentation changes must not imply unsupported runtime behavior, migration steps, or deployment requirements.

## PR Severity Rubric

1. P0: merge-blocking data loss, security exposure, corruption, broken core workflow, or irreversible operational risk.
2. P1: likely user-visible correctness, compatibility, migration, reliability, or operational failure.
3. P2: bounded edge-case defect, missing validation, observability gap, or test gap with clear impact.

## Embedded Review Discipline

### Review Scope

1. Classify each round as `first review`, `incremental review`, or `targeted verification`.
2. Use `first review` when there is no previously reviewed baseline.
3. Use `incremental review` after the first baseline; focus on the diff since that baseline and re-check prior findings only when touched by new changes.
4. Use `targeted verification` only when the round is explicitly about verifying named prior findings.

### Baseline Scan

1. Always scan in-scope changes for high-severity correctness, security, data consistency, performance, compatibility, observability, and testing risks.
2. Expert or prompt focus areas are minimum requirements, not the full scope.
3. Report only concrete bugs with plausible failure paths.

### Finding Contract

1. Order findings by severity, then confidence.
2. Each finding must include `Severity`, `Confidence`, `What`, `Evidence`, `Impact`, `Fix`, and `Verify`.
3. Severity values are `Blocker`, `High`, `Medium`, and `Low`.
4. Confidence values are `High`, `Medium`, and `Low`.
5. Do not duplicate the same finding across multiple comments.
6. Do not report speculative, style-only, or preference-only findings.
7. If no issues are found, state `No issues found` and list concrete checks performed.

### Severity Guidance

1. `Blocker`: prevents merge due to data loss, security exposure, corruption, broken core workflow, or irreversible operational risk.
2. `High`: likely user-visible correctness, concurrency, compatibility, migration, or reliability failure.
3. `Medium`: real defect with bounded impact, edge-case trigger, or missing validation that can cause incorrect behavior.
4. `Low`: non-blocking issue with clear value and low risk.

## Changes Since Previous Codex Review

There is no previous Codex review. Review the full PR diff:

1. `AGENTS.md`
   1. Expands `server/` ownership to include runtime usage.
   2. Adds global guidance that legacy API metering uses `MNEMO_METERING_*`, while runtime usage quota and console metering use `MNEMO_RUNTIME_USAGE_*` and do not use `MNEMO_METERING_URL`.
   3. Adds lookup rows for `server/internal/runtimeusage/` and `server/internal/metering/`.
2. `README.md`
   1. Adds `MNEMO_CHAIN_RECALL_STOP_SCORE` documentation.
   2. Reframes `MNEMO_METERING_*` as legacy API metering for successful recall and ingest events.
   3. Adds runtime usage quota and metering documentation for `MNEMO_RUNTIME_USAGE_*`, including base URL, secret, timeout, metering timeout, reservation TTL, operation TTL, fail-open, and outbox enabled behavior.
3. `server/AGENTS.md`
   1. Adds runtime usage quota/metering to server ownership and lookup paths.
   2. Adds guidance to keep runtime usage quota hooks at high-level handler operation boundaries.
   3. Adds guidance to keep legacy API metering separate from runtime usage quota and console metering.
   4. Adds an anti-pattern forbidding routing runtime usage console metering through `MNEMO_METERING_URL`.
4. `server/internal/handler/AGENTS.md`
   1. Adds `runtime_usage.go` as the runtime usage helper and error mapping location.
   2. Adds guidance to use existing runtime usage helpers for quota error responses and post-success finalization, including surviving request cancellation after tenant writes.
5. `server/internal/metering/AGENTS.md`
   1. Reframes the package as owning both legacy API metering and runtime usage console metering writers.
   2. Documents `ConsoleRuntimeConfig` and `NewConsoleRuntime`.
   3. Documents runtime usage console metering required fields, sanitization, and current call sites.
   4. Replaces the prior "no caller hooks yet" guidance with existing call-site guidance.

## Fixed, Rejected, Accepted, And Still-Open Findings

1. Fixed findings: none. No prior Codex or human review findings exist in the live PR state.
2. Rejected findings: none. No review threads existed for the responder to reject.
3. Accepted findings: none. No review threads existed for the responder to accept.
4. Still-open findings: none. Live review threads are empty.
5. Do-not-repeat findings: none. There are no resolved or accepted findings in this loop. If future rounds add resolved or accepted findings, do not repeat them unless the current head introduces a new failure path.

## Changed Risky Code Paths To Check

Although this PR changes documentation and agent instructions, treat these as risky because they define operational guidance and environment-variable behavior:

1. `README.md` runtime usage quota and metering table, especially whether each `MNEMO_RUNTIME_USAGE_*` default, requiredness rule, validation rule, and fail-open/outbox statement matches `server/internal/config/config.go` and runtime usage behavior.
2. `README.md` legacy metering section, especially whether it correctly describes `MNEMO_METERING_URL`, supported schemes, disabled behavior, and successful recall/ingest API event emission without confusing it with runtime usage console metering.
3. `README.md` `MNEMO_CHAIN_RECALL_STOP_SCORE`, especially whether the documented default and valid range match config validation and actual recall behavior.
4. `AGENTS.md`, `server/AGENTS.md`, `server/internal/handler/AGENTS.md`, and `server/internal/metering/AGENTS.md`, especially whether the new guidance would cause future agents to place quota hooks, metering calls, outbox usage, finalization, or environment variables in the wrong layer.
5. Cross-file consistency between root `AGENTS.md`, server AGENTS files, README, and actual code under `server/internal/config/`, `server/internal/handler/`, `server/internal/metering/`, and `server/internal/runtimeusage/`.

## Review Instructions

1. Do not load, run, or depend on second-opinion or any other external review skill.
2. Do not narrow scope based on responder conclusions. The responder made no code changes, but still verify the live PR diff independently.
3. Focus on concrete bugs only, with plausible failure paths.
4. Verify fixes, newly changed paths, cross-file regressions, and documentation-to-code mismatches.
5. Do not report speculative style preferences, wording-only nits, or duplicate findings.
6. If no issues are found, state `No issues found` and list the concrete checks performed.

File: README.md [hunk 1]
```diff
@@ -247,6 +253,24 @@ The public env surface is intentionally small for now. Metering location is conf
...
+
+| Variable | Required | Default | Description |
+|----------|----------|---------|-------------|
+| `MNEMO_RUNTIME_USAGE_ENABLED` | No | `false` | Enable runtime usage quota gating and console metering for memory recall/write operations |
+| `MNEMO_RUNTIME_USAGE_BASE_URL` | Yes when enabled | — | Runtime usage service base URL. Must be `http` or `https`; query and fragment are rejected |
+| `MNEMO_RUNTIME_USAGE_INTERNAL_SECRET` | Yes when enabled | — | Bearer token for internal runtime usage service calls |
+| `MNEMO_RUNTIME_USAGE_TIMEOUT` | No | `3s` | Timeout for quota reservation and finalization requests |
+| `MNEMO_RUNTIME_USAGE_METERING_TIMEOUT` | No | `5s` | Timeout for console metering event delivery requests |
+| `MNEMO_RUNTIME_USAGE_RESERVATION_TTL` | No | `30m` | Runtime usage reservation lifetime setting |
+| `MNEMO_RUNTIME_USAGE_OPERATION_TTL` | No | `30m` | Runtime usage operation lifetime setting |
```
[the remark this answers is not in the document]
Comment (mashenjun) [reply README.md]: Accepted and fixed in 43401937c3df0e51877446ba2a0e5e93596c06da.

`README.md` now marks both `MNEMO_RUNTIME_USAGE_RESERVATION_TTL` and `MNEMO_RUNTIME_USAGE_OPERATION_TTL` as parsed into server config but currently not behavior-changing: reservation TTL is not sent in reservation requests, and operation TTL is not used to expire runtime usage outbox rows.

**Comment by mashenjun** [comment]
@codex review

Review PR `mem9-ai/mem9#320` at current head SHA `43401937c3df0e51877446ba2a0e5e93596c06da`.

## Review Scope Classification

This is an `incremental review`.

Previous Codex review baseline: `2b6045fcea0a1d46c3b8442c4796af2c05654827`. Current PR head: `43401937c3df0e51877446ba2a0e5e93596c06da`. Focus on the incremental diff from `2b6045fcea0a1d46c3b8442c4796af2c05654827` to `43401937c3df0e51877446ba2a0e5e93596c06da`, verify the accepted prior finding was actually fixed, and check for regressions in the full changed PR surface. Do not narrow scope based on responder conclusions.

## Stable Charter Invariants

1. Runtime usage quota and console metering use `MNEMO_RUNTIME_USAGE_*`.
2. Legacy API metering uses `MNEMO_METERING_*`.
3. Runtime usage quota and console metering do not use `MNEMO_METERING_URL`.
4. Documentation must match actual environment variable behavior in the codebase.
5. Documentation changes must not imply unsupported runtime behavior, migration steps, or deployment requirements.

## PR Severity Rubric

1. P0: merge-blocking data loss, security exposure, corruption, broken core workflow, or irreversible operational risk.
2. P1: likely user-visible correctness, compatibility, migration, reliability, or operational failure.
3. P2: bounded edge-case defect, missing validation, observability gap, or test gap with clear impact.

## Embedded Review Discipline

### Review Scope

1. Classify each round as `first review`, `incremental review`, or `targeted verification`.
2. Use `first review` when there is no previously reviewed baseline.
3. Use `incremental review` after the first baseline; focus on the diff since that baseline and re-check prior findings only when touched by new changes.
4. Use `targeted verification` only when the round is explicitly about verifying named prior findings.

### Baseline Scan

1. Always scan in-scope changes for high-severity correctness, security, data consistency, performance, compatibility, observability, and testing risks.
2. Expert or prompt focus areas are minimum requirements, not the full scope.
3. Report only concrete bugs with plausible failure paths.

### Finding Contract

1. Order findings by severity, then confidence.
2. Each finding must include `Severity`, `Confidence`, `What`, `Evidence`, `Impact`, `Fix`, and `Verify`.
3. Severity values are `Blocker`, `High`, `Medium`, and `Low`.
4. Confidence values are `High`, `Medium`, and `Low`.
5. Do not duplicate the same finding across multiple comments.
6. Do not report speculative, style-only, or preference-only findings.
7. If no issues are found, state `No issues found` and list concrete checks performed.

### Severity Guidance

1. `Blocker`: prevents merge due to data loss, security exposure, corruption, broken core workflow, or irreversible operational risk.
2. `High`: likely user-visible correctness, concurrency, compatibility, migration, or reliability failure.
3. `Medium`: real defect with bounded impact, edge-case trigger, or missing validation that can cause incorrect behavior.
4. `Low`: non-blocking issue with clear value and low risk.

## Changes Since Previous Codex Review

Since the previous Codex review at `2b6045fcea0a1d46c3b8442c4796af2c05654827`, the PR is ahead by one commit:

1. `43401937c3df0e51877446ba2a0e5e93596c06da` - `docs: clarify runtime usage TTL knobs`.
2. Only `README.md` changed in that commit: 2 additions and 2 deletions.
3. The two runtime usage TTL rows were changed from behavior-changing lifetime descriptions to explicit parsed-but-currently-unused descriptions:
   1. `MNEMO_RUNTIME_USAGE_RESERVATION_TTL` is parsed into server config but currently not sent to reservation requests.
   2. `MNEMO_RUNTIME_USAGE_OPERATION_TTL` is parsed into server config but currently not used to expire runtime usage outbox rows.

Full PR changed files at current head:

1. `AGENTS.md`
2. `README.md`
3. `server/AGENTS.md`
4. `server/internal/handler/AGENTS.md`
5. `server/internal/metering/AGENTS.md`

## Fixed, Rejected, Accepted, And Still-Open Findings

1. Fixed findings:
   1. `PRRT_kwDORhSfys6DHY9j`, root comment database ID `3265499603`, Codex finding `Avoid documenting no-op runtime usage TTL knobs`, original path `README.md`, original lines 269-270.
   2. Live review thread state is `isResolved=true` and `isOutdated=true`.
   3. The fix is the README change in `43401937c3df0e51877446ba2a0e5e93596c06da` marking both TTL variables as parsed but not currently behavior-changing.
2. Rejected findings: none.
3. Accepted findings:
   1. The TTL documentation finding above was accepted as valid and fixed.
   2. Treat this accepted/resolved finding as do-not-repeat unless current head introduces a new concrete failure path, such as the README still implying either TTL changes reservation payloads or runtime usage outbox expiry.
4. Still-open findings: none. Live review threads currently show no unresolved threads.

## Changed Risky Code Paths To Check

Although this PR changes documentation and agent instructions, treat these as risky because they define operational behavior for future agents and operators:

1. `README.md` runtime usage quota and metering table versus `server/internal/config/config.go`, `server/internal/runtimeusage/client.go`, `server/internal/runtimeusage/outbox.go`, `server/internal/runtimeusage/manager.go`, and `server/internal/metering/console_writer.go`.
2. The fixed `README.md` rows for `MNEMO_RUNTIME_USAGE_RESERVATION_TTL` and `MNEMO_RUNTIME_USAGE_OPERATION_TTL`, especially whether the new wording is accurate and no longer advertises unsupported behavior.
3. `README.md` legacy metering section versus `server/internal/config/config.go`, `server/internal/handler/metering.go`, and `server/internal/metering/`, especially separation between `MNEMO_METERING_*` legacy API metering and `MNEMO_RUNTIME_USAGE_*` runtime usage console metering.
4. `README.md` `MNEMO_CHAIN_RECALL_STOP_SCORE` row versus `server/internal/config/config.go` validation and actual chain recall behavior.
5. `AGENTS.md`, `server/AGENTS.md`, `server/internal/handler/AGENTS.md`, and `server/internal/metering/AGENTS.md`, especially whether the new guidance could direct future agents to put runtime usage hooks, finalization, metering calls, outbox handling, or environment-variable usage in the wrong layer.
6. Cross-file consistency between the README, root `AGENTS.md`, server AGENTS files, and actual code under `server/internal/config/`, `server/internal/handler/`, `server/internal/metering/`, and `server/internal/runtimeusage/`.

## Review Instructions

1. Do not load, run, or depend on second-opinion or any other external review skill.
2. Verify the resolved TTL finding independently from PR evidence; do not trust the responder artifact by itself.
3. Focus on verifying fixes, newly changed code paths, full-PR changed risky paths, and regressions.
4. Report concrete bugs only, with plausible failure paths.
5. Do not repeat the resolved/accepted TTL finding unless the current head introduces a new failure path.
6. Do not report speculative style preferences, wording-only nits, or duplicate findings.
7. If no issues are found, state `No issues found` and list the concrete checks performed.
